public/app/features/explore/RichHistory/RichHistoryQueriesTab.tsx

```diff
@@ -202,13 +203,20 @@ export function RichHistoryQueriesTab(props: Props) {
           </div>
         </div>
         {Object.keys(mappedQueriesToHeadings).map(heading => {
+          if (counter === renderLimit) {
+            return null;
+          }
           return (
             <div key={heading}>
               <div className={styles.heading}>
                 {heading} <span className={styles.queries}>{mappedQueriesToHeadings[heading].length} queries</span>
               </div>
               {mappedQueriesToHeadings[heading].map((q: RichHistoryQuery) => {
+                if (counter === renderLimit) {
+                  return null;
+                }
                 const idx = listOfDatasources.findIndex(d => d.label === q.datasourceName);
+                counter++;
```

Comment:
Isn't it better (performance-wise) to work on a slice instead of manually incrementing the counter on every re-render?

Reply:
because of the way how is mappedQueriesToHeadings structured (we use it as mapping of headings (data sources names or dates), using slice is not possible (at least I couldn't think of the way how to use it - we would still have to use counter).

Reply:
Doesn't it return just an array of objects though? You can rewrite your counter-based logic to a slice easily.

Reply:
It returns object with arrays mapped to heading.

Reply:
`mappedQueriesToHeadings[heading]` is basically an array of objects, so with a given limit K, instead of iterating over N elements every time (that can go into thousands) of which N-K is basically going to be nulls (according to your mapping), you can do a simple slice on `mappedQueriesToHeadings[heading]`, e.g. like `mappedQueriesToHeadings[heading].slice(0, K)` and it should be more performant - as you do not need to iterate over 1000 elements with a counter when you want to render only 100 elements.

Reply:
I mean, with a long query history we basically do a lot of unnecessary iterations every single step. Isn't it going to be better to keep the count of queries per each heading somewhere outside of this mapping function (since the number of headings is finite and should have a hard limit of elements, e.g. the number of days) and slice the headings (and their queries) accordingly?

Reply:
Can you maybe create a PR draft for that?  Because I have thought about it a lot and I am not sure how exactly should it be implemented. We could use slice, but still we would need a counter. But maybe I am not seeing something obvious. 

> Isn't it going to be better to keep the count of queries per each heading somewhere outside of this mapping function?

Well this is tricky as well, as headings (data source name vs. date), lengths of arrays (based on time filtering and data source name filtering) can change constantly. So you have to always calculate it on fly.